File: python/125_Valid_Palindrome.py
'''
Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.
Note: For the purpose of this problem, we define empty string as valid palindrome.

Example 1:
Input: "A man, a plan, a canal: Panama"
Output: true

Example 2:
Input: "race a car"
Output: false
'''

class Solution:
    def isPalindrome(self, s: str) -> bool:
        strs = []
        for char in s:
            if char.isalnum():
                strs.append(char.lower())

        while len(strs) > 1:
            if strs.pop(0) != strs.pop():
                return False

        return True


if __name__ == "__main__":
    input_val = "A man, a plan, a canal: Panama"
    ret_val = Solution().isPalindrome(input_val)
    print(ret_val)

File: kotlin/src/125_Valid_Palindrome.kt
import java.util.*

/*
Example 1:

Input: s = "A man, a plan, a canal: Panama"
Output: true
Explanation: "amanaplanacanalpanama" is a palindrome.
*/
fun isPalindrome(s: String): Boolean {
    val cleansingStr = s.lowercase(Locale.getDefault()).filter { it.isLetter() }
    val length = cleansingStr.length
    for(i in 0 until length / 2) {
        val frontChar = cleansingStr[i]
        val backChar = cleansingStr[length - i - 1]
        if (frontChar != backChar)
            return false
    }
    return true
}

fun main() {
    val str = "A man, a plan, a canal: Panama"
    val result = isPalindrome(str)
    print(result)
}
